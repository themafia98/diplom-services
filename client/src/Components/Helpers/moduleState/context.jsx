import { createContext } from 'react';

const ModuleContextState = createContext();
const ModuleContextActions = createContext();

export { ModuleContextState, ModuleContextActions };
