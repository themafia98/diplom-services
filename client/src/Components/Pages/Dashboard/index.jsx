import React from "react";
import PropTypes from "prop-types";
import { Redirect } from "react-router-dom";
import { EventEmitter } from "events";
import _ from "lodash";
import config from "../../../config.json";
import { Layout, message, notification, Modal, Button } from "antd";
import { connect } from "react-redux";
import {
    addTabAction,
    setActiveTabAction,
    removeTabAction,
    logoutAction,
    shouldUpdateAction
} from "../../../Redux/actions/routerActions";
import { loadCurrentData } from "../../../Redux/actions/routerActions/middleware";
import { errorRequstAction } from "../../../Redux/actions/publicActions";
import { routeParser } from "../../../Utils";

import Loader from "../../Loader";
import HeaderView from "../../HeaderView";
import ContentView from "../../ContentView";
import MenuView from "../../MenuView";

class Dashboard extends React.PureComponent {
    dashboardStrem = new EventEmitter();
    deferredPrompt = null;
    state = {
        collapsed: true,
        guideVisible: true,
        visibleInstallApp: false,
        redirect: false,
        status: "online",
        menuItems: config.menu,
        counterError: 0,
        showLoader: false
    };

    static propTypes = {
        addTab: PropTypes.func.isRequired,
        removeTab: PropTypes.func.isRequired,
        setCurrentTab: PropTypes.func.isRequired,
        onLoadCurrentData: PropTypes.func.isRequired,
        onErrorRequstAction: PropTypes.func.isRequired,
        onLogoutAction: PropTypes.func.isRequired,
        router: PropTypes.object.isRequired,
        publicReducer: PropTypes.object.isRequired
    };

    componentDidMount = () => {
        // window.addEventListener("beforeinstallprompt", e => {
        //     debugger;
        //     // Prevent Chrome 67 and earlier from automatically showing the prompt
        //     e.preventDefault();
        //     // Stash the event so it can be triggered later.
        //     this.deferredPrompt = e;
        //     console.log("install");
        //     this.setState({
        //         visibleInstallApp: true
        //     });
        // });

        window.addEventListener("appinstalled", evt => {
            console.log("appinstalled fired", evt);
        });
    };

    componentDidUpdate = () => {
        const {
            publicReducer: { requestError = null, status } = {},
            router,
            firebase,
            router: { currentActionTab = "" }
        } = this.props;
        const { showLoader, status: statusState, counterError, redirect = false } = this.state;

        if (redirect) return;
        if (!firebase.getCurrentUser()) {
            return this.setState({
                redirect: true
            });
        }

        if (!showLoader && statusState !== status && status === "online")
            notification.success({ message: "Удачно", description: "Интернет соединение восстановлено." });

        if (showLoader && _.isNull(requestError) && status === "online") {
            const { routeData = {} } = router;
            const copyRouteData = { ...routeData };
            let currentArray = currentActionTab.split("_" || "__");
            let regExp = new RegExp(currentArray[0], "gi");
            let keys = Object.keys(copyRouteData).filter(key => /Module/gi.test(key) && regExp.test(key));

            if (keys.every(key => copyRouteData[key].load === true) || requestError === "Network error") {
                return setTimeout(() => {
                    this.setState({
                        status: status,
                        showLoader: false
                    });
                }, 500);
            } else if (_.isNull(requestError) && status === "online") {
                return setTimeout(() => {
                    this.setState({
                        counter: 0,
                        status: status,
                        showLoader: false
                    });
                }, 500);
            }
        } else if (showLoader && status === "offline") {
            return setTimeout(() => {
                this.setState({
                    status: status,
                    showLoader: false
                });
            }, 500);
        }

        if (
            !_.isNull(requestError) &&
            requestError[requestError.length - 1] === "Network error" &&
            status === "offline" &&
            counterError === 0
        ) {
            this.setState({ counterError: counterError + 1 }, () =>
                notification.error({ message: "Ошибка", description: "Интернет соединение недоступно." })
            );
        }

        if (statusState !== status) {
            return this.setState({
                status: status
            });
        }
    };

    onCollapse = collapsed => {
        this.setState({ ...this.state, collapsed });
    };

    logout = event => {
        const { firebase, onLogoutAction } = this.props;
        if (firebase) firebase.signOut().then(() => onLogoutAction().then(() => this.props.history.push("/")));
    };

    closeGuild = event => {
        this.setState({
            guideVisible: false
        });
    };

    updateLoader = event => {
        const {
            router,
            router: { currentActionTab, shouldUpdate }
        } = this.props;
        const { routeData = {} } = router;
        const copyRouteData = { ...routeData };
        let currentArray = currentActionTab.split("_" || "__");
        let regExp = new RegExp(currentArray[0], "gi");
        let keys = Object.keys(copyRouteData).filter(key => /Module/gi.test(key) && regExp.test(key));

        if (keys.length && !shouldUpdate) this.setState({ showLoader: true });
    };

    getActionTabs = (tabs = [], menu) => {
        const { router: { routeData = {} } = {} } = this.props;
        const tabsCopy = [...tabs];
        const tabsArray = [];
        for (let i = 0; i < tabsCopy.length; i++) {
            let tabItem = menu.find(tab => tab.EUID === tabsCopy[i]);
            if (!tabItem) {
                const PAGEDATA = routeParser({ pageType: "page", path: tabsCopy[i] });
                const PARENT_CODE = PAGEDATA["page"];
                const DATAKEY = PAGEDATA["pageChild"] || PAGEDATA["page"] || "";
                const { listdata: { title: titleListdata = "" } = {}, title = "", name = "" } =
                    routeData[DATAKEY] || {};

                const VALUE = DATAKEY && name ? name : title ? title : titleListdata ? titleListdata : DATAKEY;
                tabItem = { EUID: PAGEDATA["path"] || tabsCopy[i], PARENT_CODE, DATAKEY, VALUE };
            }
            if (tabItem) tabsArray.push({ ...tabItem });
        }
        return tabsArray;
    };

    goHome = event => {
        const { addTab, setCurrentTab, router: { currentActionTab = "", actionTabs = [] } = {} } = this.props;
        if (currentActionTab === "mainModule") return;

        if (config.tabsLimit <= actionTabs.length)
            return message.error(`Максимальное количество вкладок: ${config.tabsLimit}`);

        let tabItem = actionTabs.find(tab => tab === "mainModule");
        if (!tabItem) addTab("mainModule");
        else setCurrentTab("mainModule");
    };

    goCabinet = event => {
        const { addTab, setCurrentTab, router: { currentActionTab = "", actionTabs = [] } = {} } = this.props;

        if (currentActionTab === "cabinetModule") return;

        if (config.tabsLimit <= actionTabs.length)
            return message.error(`Максимальное количество вкладок: ${config.tabsLimit}`);

        let tabItem = actionTabs.find(tab => tab === "cabinetModule");
        if (!tabItem) addTab("cabinetModule");
        else setCurrentTab("cabinetModule");
    };

    menuHandler = (event, key, mode = "open") => {
        const path = event["key"] ? event["key"] : key;
        const { router: { currentActionTab, actionTabs = [] } = {}, addTab, setCurrentTab, removeTab } = this.props;

        const actionTabsCopy = [...actionTabs];
        const isFind = actionTabsCopy.findIndex(tab => tab === path) !== -1;

        if (mode === "open") {
            if (!isFind && config.tabsLimit <= actionTabsCopy.length)
                return message.error(`Максимальное количество вкладок: ${config.tabsLimit}`);

            if (!isFind) addTab(routeParser({ path }));
            else if (currentActionTab !== path) {
                setCurrentTab(path);
            }
        } else if (mode === "close") {
            let type = "deafult";
            if (path.split("__")[1]) type = "itemTab";
            if (isFind) removeTab({ path: path, type: type });
        }
    };

    installApp = event => {
        if (!this.deferredPrompt) {
            return notification.error({
                message: "Ошибка",
                description: "Невозможно скачать приложение, возможно ваш браузер устарел."
            });
        }
        this.deferredPrompt.prompt();

        this.deferredPrompt.userChoice.then(choiceResult => {
            if (choiceResult.outcome === "accepted") {
                console.log("PWA setup accepted");
                // hide our user interface that shows our A2HS button
            } else {
                console.log("PWA setup rejected");
            }
            this.deferredPrompt = null;
        });
    };

    render() {
        const { menuItems = null, showLoader, redirect, guideVisible, visibleInstallApp = false } = this.state;
        const {
            router: { actionTabs = [], currentActionTab, shouldUpdate = false } = {},
            router = {},
            firebase,
            onErrorRequstAction,
            onShoudUpdate,
            firstConnect = false,
            setCurrentTab
        } = this.props;

        if (redirect) return <Redirect to={{ pathname: "/" }} />;

        const actionTabsData = this.getActionTabs(actionTabs, menuItems);

        return (
            <React.Fragment>
                {showLoader ? <Loader className="mainLoader" /> : null}
                <Layout className="layout_menu">
                    <MenuView
                        key="menu"
                        items={menuItems}
                        activeTabEUID={currentActionTab}
                        cbMenuHandler={this.menuHandler}
                        collapsed={this.state.collapsed}
                        cbOnCollapse={this.onCollapse}
                        cbGoMain={this.goHome}
                    />
                    <Layout key="main">
                        <HeaderView
                            key="header"
                            dashboardStrem={this.dashboardStrem}
                            cbMenuTabHandler={this.menuHandler}
                            activeTabEUID={currentActionTab}
                            actionTabs={actionTabsData ? actionTabsData : false}
                            logout={this.logout}
                            goCabinet={this.goCabinet}
                        />
                        <ContentView
                            dashboardStrem={this.dashboardStrem}
                            actionTabs={actionTabs}
                            shouldUpdate={shouldUpdate}
                            router={router}
                            onShoudUpdate={onShoudUpdate}
                            setCurrentTab={setCurrentTab}
                            updateLoader={this.updateLoader}
                            onErrorRequstAction={onErrorRequstAction}
                            firebase={firebase}
                            key="contentView"
                            path={currentActionTab}
                        />
                        <Modal
                            className="guideModal"
                            okText="Закрыть"
                            title="Добро пожаловать в систему."
                            visible={guideVisible && firstConnect}
                            onOk={this.closeGuild}
                        >
                            <p>Спасибо что выбрали {config.title ? `${config.title}` : "нашу систему!"}</p>
                            {visibleInstallApp ? (
                                <React.Fragment>
                                    <p>Вы можете установить приложение на ваш компьютер</p>

                                    <Button onClick={this.installApp} className="setupButton">
                                        Установить приложение
                                    </Button>
                                </React.Fragment>
                            ) : null}
                        </Modal>
                    </Layout>
                </Layout>
            </React.Fragment>
        );
    }
}

const mapStateToProps = state => {
    const { firstConnect = false } = state.publicReducer;
    return {
        router: { ...state.router },
        publicReducer: state.publicReducer,
        firstConnect
    };
};

const mapDispatchToProps = dispatch => {
    return {
        addTab: tab => dispatch(addTabAction(tab)),
        removeTab: tab => dispatch(removeTabAction(tab)),
        setCurrentTab: tab => dispatch(setActiveTabAction(tab)),
        onLoadCurrentData: ({ path, storeLoad }) => dispatch(loadCurrentData({ path, storeLoad })),
        onErrorRequstAction: async error => await errorRequstAction(error),
        onShoudUpdate: async update => await shouldUpdateAction(update),
        onLogoutAction: async () => await dispatch(logoutAction())
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
export { Dashboard };
