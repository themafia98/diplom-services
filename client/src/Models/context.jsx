import { createContext } from 'react';
import config from '../config.json';
import Request from '../Models/Rest';
import Schema from '../Models/Schema';
import Notification from '../Models/NotificationRouter';
import { clientDB } from './ClientSideDatabase';

export const modelMethods = {
  clientDB,
  rest: new Request(),
  Request,
  config,
  schemaModel: Schema,
  schema: new Schema('no-strict'),
  Notification,
};

const modelsContext = createContext(modelMethods);

export default modelsContext;
