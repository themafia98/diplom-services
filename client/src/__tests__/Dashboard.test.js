// @ts-nocheck
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Dashboard } from '../Pages/Dashboard/Dashboard';
import context from '../Models/context';
describe('<Dashboard /> main page', () => {
  test('Should render component', () => {
    const props = {
      addTab: () => {},
      removeTab: () => {},
      setCurrentTab: () => {},
      onLoadCurrentData: () => {},
      onErrorRequestAction: () => {},
      onLogoutAction: () => {},
      publicReducer: {
        status: 'online',
        prewStatus: 'online',
        requestError: null,
        caches: {},
      },
      router: {
        currentActionTab: 'MainModule',
        actionTabs: ['MainModule'],
        routeDataActive: null,
        routeData: { MainModule: { name: 'MainModule' } },
      },
      rest: {},
      firstConnect: false,
      udata: { _id: 'dasd', displayName: 'asdas' },
      onSetStatus: () => {},
      onClearCache: () => {},
      onShoudUpdate: () => {},
      onShowGuide: () => {},
      tabData: {},
    };
    const DashboardWrapper = shallow(<Dashboard {...props} />, { context });

    expect(DashboardWrapper.find('MenuView').prop('collapsed')).toEqual(true);
    expect(DashboardWrapper.find('ContentView').prop('path')).toEqual(props.router.currentActionTab);

    DashboardWrapper.setState({ collapsed: true });
    expect(DashboardWrapper.find('MenuView').prop('collapsed')).toBeTruthy();

    DashboardWrapper.setState({ collapsed: false });
    expect(DashboardWrapper.find('MenuView').prop('collapsed')).toEqual(false);

    DashboardWrapper.setState({ showLoader: true });
    expect(toJson(DashboardWrapper)).toMatchSnapshot();

    DashboardWrapper.setState({ showLoader: false });
    expect(toJson(DashboardWrapper)).toMatchSnapshot();

    DashboardWrapper.setState({ menuItems: [] });
    expect(toJson(DashboardWrapper)).toMatchSnapshot();
  });
});
