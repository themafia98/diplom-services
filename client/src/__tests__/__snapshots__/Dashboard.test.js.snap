// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> main page Should render component 1`] = `
<div
  className="dashboard"
>
  <Loader
    className="mainLoader"
    classNameSpiner={null}
  />
  <Layout
    className="layout_menu"
  >
    <MenuView
      activeTabEUID="MainModule"
      cbGoMain={[Function]}
      cbMenuHandler={[Function]}
      cbOnCollapse={[Function]}
      collapsed={false}
      items={null}
      key="menu"
    />
    <Layout
      key="main"
    >
      <Connect(HeaderView)
        activeTabs={
          Array [
            Object {
              "DATAKEY": "MainModule",
              "EUID": "MainModule",
              "PARENT_CODE": "MainModule",
              "VALUE": "MainModule",
            },
          ]
        }
        activeTabEUID="MainModule"
        cbMenuTabHandler={[Function]}
        dashboardStrem={
          EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            Symbol(kCapture): false,
          }
        }
        goCabinet={[Function]}
        key="header"
        logout={[Function]}
      />
      <ContentView
        activeTabs={
          Array [
            "MainModule",
          ]
        }
        dashboardStrem={
          EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            Symbol(kCapture): false,
          }
        }
        isToolbarActive={false}
        key="contentView"
        onChangeVisibleAction={[Function]}
        onErrorRequestAction={[Function]}
        onHideLoader={[Function]}
        onSetStatus={[Function]}
        onShoudUpdate={[Function]}
        onShowLoader={[Function]}
        path="MainModule"
        rest={Object {}}
        router={
          Object {
            "activeTabs": Array [
              "MainModule",
            ],
            "currentActionTab": "MainModule",
            "routeData": Object {
              "MainModule": Object {
                "name": "MainModule",
              },
            },
            "routeDataActive": null,
          }
        }
        setCurrentTab={[Function]}
        shouldRenderMenu={true}
        shouldUpdate={false}
        statusApp="online"
        udata={
          Object {
            "_id": "dasd",
            "displayName": "asdas",
          }
        }
        updateLoader={[Function]}
        visibilityPortal={false}
        webSocket={
          Socket {
            "_callbacks": Object {
              "$connect": Array [
                [Function],
              ],
              "$connecting": Array [
                [Function],
              ],
            },
            "acks": Object {},
            "connected": false,
            "disconnected": true,
            "flags": Object {},
            "ids": 0,
            "io": Manager {
              "_callbacks": Object {
                "$close": Array [
                  [Function],
                ],
                "$open": Array [
                  [Function],
                ],
                "$packet": Array [
                  [Function],
                ],
              },
              "_randomizationFactor": 0.5,
              "_reconnection": true,
              "_reconnectionAttempts": Infinity,
              "_reconnectionDelay": 1000,
              "_reconnectionDelayMax": 5000,
              "_timeout": 20000,
              "autoConnect": true,
              "backoff": Backoff {
                "attempts": 0,
                "factor": 2,
                "jitter": 0.5,
                "max": 5000,
                "ms": 1000,
              },
              "connecting": Array [
                [Circular],
              ],
              "decoder": Decoder {
                "reconstructor": null,
              },
              "encoder": Encoder {},
              "encoding": false,
              "engine": Socket {
                "_callbacks": Object {
                  "$error": Array [
                    [Function],
                  ],
                  "$open": Array [
                    [Function],
                  ],
                },
                "agent": false,
                "binaryType": null,
                "ca": null,
                "cert": null,
                "ciphers": null,
                "enablesXDR": false,
                "forceBase64": false,
                "forceJSONP": false,
                "forceNode": false,
                "hostname": "localhost",
                "id": null,
                "isReactNative": false,
                "jsonp": true,
                "key": null,
                "onlyBinaryUpgrades": undefined,
                "passphrase": null,
                "path": "/socket.io/",
                "perMessageDeflate": Object {
                  "threshold": 1024,
                },
                "pfx": null,
                "pingInterval": null,
                "pingIntervalTimer": null,
                "pingTimeout": null,
                "pingTimeoutTimer": null,
                "policyPort": 843,
                "port": "80",
                "prevBufferLen": 0,
                "query": Object {},
                "readyState": "opening",
                "rejectUnauthorized": true,
                "rememberUpgrade": false,
                "secure": false,
                "timestampParam": "t",
                "timestampRequests": undefined,
                "transport": WS {
                  "_callbacks": Object {
                    "$close": Array [
                      [Function],
                    ],
                    "$drain": Array [
                      [Function],
                    ],
                    "$error": Array [
                      [Function],
                    ],
                    "$packet": Array [
                      [Function],
                    ],
                  },
                  "agent": false,
                  "ca": null,
                  "cert": null,
                  "ciphers": null,
                  "enablesXDR": false,
                  "extraHeaders": undefined,
                  "forceNode": false,
                  "hostname": "localhost",
                  "isReactNative": false,
                  "key": null,
                  "localAddress": undefined,
                  "passphrase": null,
                  "path": "/socket.io/",
                  "perMessageDeflate": Object {
                    "threshold": 1024,
                  },
                  "pfx": null,
                  "port": "80",
                  "protocols": undefined,
                  "query": Object {
                    "EIO": 3,
                    "transport": "websocket",
                  },
                  "readyState": "opening",
                  "rejectUnauthorized": true,
                  "secure": false,
                  "socket": [Circular],
                  "timestampParam": "t",
                  "timestampRequests": undefined,
                  "usingBrowserWebSocket": true,
                  "withCredentials": true,
                  "ws": WebSocket {},
                },
                "transportOptions": Object {},
                "transports": Array [
                  "websocket",
                ],
                "upgrade": true,
                "upgrades": null,
                "withCredentials": true,
                "writeBuffer": Array [],
              },
              "lastPing": null,
              "nsps": Object {
                "/": [Circular],
              },
              "opts": Object {
                "hostname": "localhost",
                "path": "/socket.io",
                "port": "80",
                "secure": false,
                "transports": Array [
                  "websocket",
                ],
              },
              "packetBuffer": Array [],
              "readyState": "opening",
              "skipReconnect": false,
              "subs": Array [
                Object {
                  "destroy": [Function],
                },
                Object {
                  "destroy": [Function],
                },
                Object {
                  "destroy": [Function],
                },
              ],
              "uri": "http://localhost/",
            },
            "json": [Circular],
            "nsp": "/",
            "receiveBuffer": Array [],
            "sendBuffer": Array [],
            "subs": Array [
              Object {
                "destroy": [Function],
              },
              Object {
                "destroy": [Function],
              },
              Object {
                "destroy": [Function],
              },
            ],
          }
        }
      />
    </Layout>
  </Layout>
  <ForwardRef
    customRender={null}
    key="toolbar"
    name="Чат"
    onChangeVisibleAction={[Function]}
  />
</div>
`;

exports[`<Dashboard /> main page Should render component 2`] = `
<div
  className="dashboard"
>
  <Layout
    className="layout_menu"
  >
    <MenuView
      activeTabEUID="MainModule"
      cbGoMain={[Function]}
      cbMenuHandler={[Function]}
      cbOnCollapse={[Function]}
      collapsed={false}
      items={null}
      key="menu"
    />
    <Layout
      key="main"
    >
      <Connect(HeaderView)
        activeTabs={
          Array [
            Object {
              "DATAKEY": "MainModule",
              "EUID": "MainModule",
              "PARENT_CODE": "MainModule",
              "VALUE": "MainModule",
            },
          ]
        }
        activeTabEUID="MainModule"
        cbMenuTabHandler={[Function]}
        dashboardStrem={
          EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            Symbol(kCapture): false,
          }
        }
        goCabinet={[Function]}
        key="header"
        logout={[Function]}
      />
      <ContentView
        activeTabs={
          Array [
            "MainModule",
          ]
        }
        dashboardStrem={
          EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            Symbol(kCapture): false,
          }
        }
        isToolbarActive={false}
        key="contentView"
        onChangeVisibleAction={[Function]}
        onErrorRequestAction={[Function]}
        onHideLoader={[Function]}
        onSetStatus={[Function]}
        onShoudUpdate={[Function]}
        onShowLoader={[Function]}
        path="MainModule"
        rest={Object {}}
        router={
          Object {
            "activeTabs": Array [
              "MainModule",
            ],
            "currentActionTab": "MainModule",
            "routeData": Object {
              "MainModule": Object {
                "name": "MainModule",
              },
            },
            "routeDataActive": null,
          }
        }
        setCurrentTab={[Function]}
        shouldRenderMenu={true}
        shouldUpdate={false}
        statusApp="online"
        udata={
          Object {
            "_id": "dasd",
            "displayName": "asdas",
          }
        }
        updateLoader={[Function]}
        visibilityPortal={false}
        webSocket={
          Socket {
            "_callbacks": Object {
              "$connect": Array [
                [Function],
              ],
              "$connecting": Array [
                [Function],
              ],
            },
            "acks": Object {},
            "connected": false,
            "disconnected": true,
            "flags": Object {},
            "ids": 0,
            "io": Manager {
              "_callbacks": Object {
                "$close": Array [
                  [Function],
                ],
                "$open": Array [
                  [Function],
                ],
                "$packet": Array [
                  [Function],
                ],
              },
              "_randomizationFactor": 0.5,
              "_reconnection": true,
              "_reconnectionAttempts": Infinity,
              "_reconnectionDelay": 1000,
              "_reconnectionDelayMax": 5000,
              "_timeout": 20000,
              "autoConnect": true,
              "backoff": Backoff {
                "attempts": 0,
                "factor": 2,
                "jitter": 0.5,
                "max": 5000,
                "ms": 1000,
              },
              "connecting": Array [
                [Circular],
              ],
              "decoder": Decoder {
                "reconstructor": null,
              },
              "encoder": Encoder {},
              "encoding": false,
              "engine": Socket {
                "_callbacks": Object {
                  "$error": Array [
                    [Function],
                  ],
                  "$open": Array [
                    [Function],
                  ],
                },
                "agent": false,
                "binaryType": null,
                "ca": null,
                "cert": null,
                "ciphers": null,
                "enablesXDR": false,
                "forceBase64": false,
                "forceJSONP": false,
                "forceNode": false,
                "hostname": "localhost",
                "id": null,
                "isReactNative": false,
                "jsonp": true,
                "key": null,
                "onlyBinaryUpgrades": undefined,
                "passphrase": null,
                "path": "/socket.io/",
                "perMessageDeflate": Object {
                  "threshold": 1024,
                },
                "pfx": null,
                "pingInterval": null,
                "pingIntervalTimer": null,
                "pingTimeout": null,
                "pingTimeoutTimer": null,
                "policyPort": 843,
                "port": "80",
                "prevBufferLen": 0,
                "query": Object {},
                "readyState": "opening",
                "rejectUnauthorized": true,
                "rememberUpgrade": false,
                "secure": false,
                "timestampParam": "t",
                "timestampRequests": undefined,
                "transport": WS {
                  "_callbacks": Object {
                    "$close": Array [
                      [Function],
                    ],
                    "$drain": Array [
                      [Function],
                    ],
                    "$error": Array [
                      [Function],
                    ],
                    "$packet": Array [
                      [Function],
                    ],
                  },
                  "agent": false,
                  "ca": null,
                  "cert": null,
                  "ciphers": null,
                  "enablesXDR": false,
                  "extraHeaders": undefined,
                  "forceNode": false,
                  "hostname": "localhost",
                  "isReactNative": false,
                  "key": null,
                  "localAddress": undefined,
                  "passphrase": null,
                  "path": "/socket.io/",
                  "perMessageDeflate": Object {
                    "threshold": 1024,
                  },
                  "pfx": null,
                  "port": "80",
                  "protocols": undefined,
                  "query": Object {
                    "EIO": 3,
                    "transport": "websocket",
                  },
                  "readyState": "opening",
                  "rejectUnauthorized": true,
                  "secure": false,
                  "socket": [Circular],
                  "timestampParam": "t",
                  "timestampRequests": undefined,
                  "usingBrowserWebSocket": true,
                  "withCredentials": true,
                  "ws": WebSocket {},
                },
                "transportOptions": Object {},
                "transports": Array [
                  "websocket",
                ],
                "upgrade": true,
                "upgrades": null,
                "withCredentials": true,
                "writeBuffer": Array [],
              },
              "lastPing": null,
              "nsps": Object {
                "/": [Circular],
              },
              "opts": Object {
                "hostname": "localhost",
                "path": "/socket.io",
                "port": "80",
                "secure": false,
                "transports": Array [
                  "websocket",
                ],
              },
              "packetBuffer": Array [],
              "readyState": "opening",
              "skipReconnect": false,
              "subs": Array [
                Object {
                  "destroy": [Function],
                },
                Object {
                  "destroy": [Function],
                },
                Object {
                  "destroy": [Function],
                },
              ],
              "uri": "http://localhost/",
            },
            "json": [Circular],
            "nsp": "/",
            "receiveBuffer": Array [],
            "sendBuffer": Array [],
            "subs": Array [
              Object {
                "destroy": [Function],
              },
              Object {
                "destroy": [Function],
              },
              Object {
                "destroy": [Function],
              },
            ],
          }
        }
      />
    </Layout>
  </Layout>
  <ForwardRef
    customRender={null}
    key="toolbar"
    name="Чат"
    onChangeVisibleAction={[Function]}
  />
</div>
`;

exports[`<Dashboard /> main page Should render component 3`] = `
<div
  className="dashboard"
>
  <Layout
    className="layout_menu"
  >
    <MenuView
      activeTabEUID="MainModule"
      cbGoMain={[Function]}
      cbMenuHandler={[Function]}
      cbOnCollapse={[Function]}
      collapsed={false}
      items={Array []}
      key="menu"
    />
    <Layout
      key="main"
    >
      <Connect(HeaderView)
        activeTabs={
          Array [
            Object {
              "DATAKEY": "MainModule",
              "EUID": "MainModule",
              "PARENT_CODE": "MainModule",
              "VALUE": "MainModule",
            },
          ]
        }
        activeTabEUID="MainModule"
        cbMenuTabHandler={[Function]}
        dashboardStrem={
          EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            Symbol(kCapture): false,
          }
        }
        goCabinet={[Function]}
        key="header"
        logout={[Function]}
      />
      <ContentView
        activeTabs={
          Array [
            "MainModule",
          ]
        }
        dashboardStrem={
          EventEmitter {
            "_events": Object {},
            "_eventsCount": 0,
            "_maxListeners": undefined,
            Symbol(kCapture): false,
          }
        }
        isToolbarActive={false}
        key="contentView"
        onChangeVisibleAction={[Function]}
        onErrorRequestAction={[Function]}
        onHideLoader={[Function]}
        onSetStatus={[Function]}
        onShoudUpdate={[Function]}
        onShowLoader={[Function]}
        path="MainModule"
        rest={Object {}}
        router={
          Object {
            "activeTabs": Array [
              "MainModule",
            ],
            "currentActionTab": "MainModule",
            "routeData": Object {
              "MainModule": Object {
                "name": "MainModule",
              },
            },
            "routeDataActive": null,
          }
        }
        setCurrentTab={[Function]}
        shouldRenderMenu={true}
        shouldUpdate={false}
        statusApp="online"
        udata={
          Object {
            "_id": "dasd",
            "displayName": "asdas",
          }
        }
        updateLoader={[Function]}
        visibilityPortal={false}
        webSocket={
          Socket {
            "_callbacks": Object {
              "$connect": Array [
                [Function],
              ],
              "$connecting": Array [
                [Function],
              ],
            },
            "acks": Object {},
            "connected": false,
            "disconnected": true,
            "flags": Object {},
            "ids": 0,
            "io": Manager {
              "_callbacks": Object {
                "$close": Array [
                  [Function],
                ],
                "$open": Array [
                  [Function],
                ],
                "$packet": Array [
                  [Function],
                ],
              },
              "_randomizationFactor": 0.5,
              "_reconnection": true,
              "_reconnectionAttempts": Infinity,
              "_reconnectionDelay": 1000,
              "_reconnectionDelayMax": 5000,
              "_timeout": 20000,
              "autoConnect": true,
              "backoff": Backoff {
                "attempts": 0,
                "factor": 2,
                "jitter": 0.5,
                "max": 5000,
                "ms": 1000,
              },
              "connecting": Array [
                [Circular],
              ],
              "decoder": Decoder {
                "reconstructor": null,
              },
              "encoder": Encoder {},
              "encoding": false,
              "engine": Socket {
                "_callbacks": Object {
                  "$error": Array [
                    [Function],
                  ],
                  "$open": Array [
                    [Function],
                  ],
                },
                "agent": false,
                "binaryType": null,
                "ca": null,
                "cert": null,
                "ciphers": null,
                "enablesXDR": false,
                "forceBase64": false,
                "forceJSONP": false,
                "forceNode": false,
                "hostname": "localhost",
                "id": null,
                "isReactNative": false,
                "jsonp": true,
                "key": null,
                "onlyBinaryUpgrades": undefined,
                "passphrase": null,
                "path": "/socket.io/",
                "perMessageDeflate": Object {
                  "threshold": 1024,
                },
                "pfx": null,
                "pingInterval": null,
                "pingIntervalTimer": null,
                "pingTimeout": null,
                "pingTimeoutTimer": null,
                "policyPort": 843,
                "port": "80",
                "prevBufferLen": 0,
                "query": Object {},
                "readyState": "opening",
                "rejectUnauthorized": true,
                "rememberUpgrade": false,
                "secure": false,
                "timestampParam": "t",
                "timestampRequests": undefined,
                "transport": WS {
                  "_callbacks": Object {
                    "$close": Array [
                      [Function],
                    ],
                    "$drain": Array [
                      [Function],
                    ],
                    "$error": Array [
                      [Function],
                    ],
                    "$packet": Array [
                      [Function],
                    ],
                  },
                  "agent": false,
                  "ca": null,
                  "cert": null,
                  "ciphers": null,
                  "enablesXDR": false,
                  "extraHeaders": undefined,
                  "forceNode": false,
                  "hostname": "localhost",
                  "isReactNative": false,
                  "key": null,
                  "localAddress": undefined,
                  "passphrase": null,
                  "path": "/socket.io/",
                  "perMessageDeflate": Object {
                    "threshold": 1024,
                  },
                  "pfx": null,
                  "port": "80",
                  "protocols": undefined,
                  "query": Object {
                    "EIO": 3,
                    "transport": "websocket",
                  },
                  "readyState": "opening",
                  "rejectUnauthorized": true,
                  "secure": false,
                  "socket": [Circular],
                  "timestampParam": "t",
                  "timestampRequests": undefined,
                  "usingBrowserWebSocket": true,
                  "withCredentials": true,
                  "ws": WebSocket {},
                },
                "transportOptions": Object {},
                "transports": Array [
                  "websocket",
                ],
                "upgrade": true,
                "upgrades": null,
                "withCredentials": true,
                "writeBuffer": Array [],
              },
              "lastPing": null,
              "nsps": Object {
                "/": [Circular],
              },
              "opts": Object {
                "hostname": "localhost",
                "path": "/socket.io",
                "port": "80",
                "secure": false,
                "transports": Array [
                  "websocket",
                ],
              },
              "packetBuffer": Array [],
              "readyState": "opening",
              "skipReconnect": false,
              "subs": Array [
                Object {
                  "destroy": [Function],
                },
                Object {
                  "destroy": [Function],
                },
                Object {
                  "destroy": [Function],
                },
              ],
              "uri": "http://localhost/",
            },
            "json": [Circular],
            "nsp": "/",
            "receiveBuffer": Array [],
            "sendBuffer": Array [],
            "subs": Array [
              Object {
                "destroy": [Function],
              },
              Object {
                "destroy": [Function],
              },
              Object {
                "destroy": [Function],
              },
            ],
          }
        }
      />
    </Layout>
  </Layout>
  <ForwardRef
    customRender={null}
    key="toolbar"
    name="Чат"
    onChangeVisibleAction={[Function]}
  />
</div>
`;
